File: pistoleiros/salao/salao.py
import sys
import time

mapSize = int(sys.argv[1])
b = int(sys.argv[2])
c = int(sys.argv[3])


tabuleiro = [["_" for _ in range(mapSize)] for _ in range(mapSize)]
tabuleiros = set()


def salao(b, c, map, solutions,first):
    if b == 0 and c == 0:
        if checaTabuleiro(map):
            solutions.add(tuple(tuple(row) for row in map))
            for i in range(4):
                map = rotaTabuleiro(map)
                solutions.add(tuple(tuple(row) for row in map))
            """ for row in map:
                print(" ".join(row)) """
        return

    if b < 0 or c < 0:
        return



   
    #primeiro b
    if first:
        for i in range((len(map)+1)//2):
            for j in range((len(map)+1)//2):
                if b>0 and notContiguous(i, j, map):
                    map[i][j] = "b"
                    b -= 1
                    salao(b, c, map, solutions,False)
                    b += 1
                    map[i][j] = "_"
                   
       

    # Primeira adição de b
    for i in range(len(map)):
        for j in range(len(map)):
            if map[i][j] =="_" and notContiguous(i, j, map) and b > 0:
                map[i][j] = "b"
                b -= 1
                salao(b, c, map, solutions,False)
                b += 1
                map[i][j] = "_"


    # Adição de c
    for i in range(len(map)):
        for j in range(len(map)):
            if map[i][j]=="_" and isValid(i, j, map, "c", "b") and c > 0:
                c -= 1
                map[i][j] = "c"
                salao(b, c, map, solutions,False)
                c += 1
                map[i][j] = "_"


    return solutions


def notContiguous(l, c, map):
    # check all directions
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1),
                  (-1, -1), (-1, 1), (1, -1), (1, 1)]
    for dr, dc in directions:
        if 0 <= l + dr < len(map) and 0 <= c + dc < len(map):
            if map[l + dr][c + dc] == "b":
                return False
    return True


def isValid(l, c, map, pistoleiro, enemy):
    enemySeen = 0
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1),
                  (-1, -1), (-1, 1), (1, -1), (1, 1)]
    for dr, dc in directions:
        for step in range(1, mapSize):
            r, d = l + dr * step, c + dc * step
            if not (0 <= r < mapSize and 0 <= d < mapSize):
                break
            if map[r][d] == pistoleiro:
                return False
            if map[r][d] == enemy:
                enemySeen += 1
                break


    return enemySeen >= 2


def checaTabuleiro(map):
    for i in range(len(map)):
        for j in range(len(map)):
            if map[i][j] == "b":
                if not isValid(i, j, map, "b", "c"):
                    return False
    return True




def rotaTabuleiro(tabuleiro):
    size = len(tabuleiro)
    rotado = [[0] * size for _ in range(size)]
    for i in range(size):
        for j in range(size):
            rotado[j][size-1-i] = tabuleiro[i][j]
    return rotado



inicio = time.time() 
print(len(salao(b, c, tabuleiro, tabuleiros,True)))
fim = time.time() 
tempo_execucao = fim - inicio
print(f"Tempo de execução: {tempo_execucao:.6f} segundos")

File: pistoleiros/salao/salao.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

int mapSize;
int b;
int c;

char **createBoard(int size) {
    char **board = (char **)malloc(size * sizeof(char *));
    for (int i = 0; i < size; i++) {
        board[i] = (char *)malloc(size * sizeof(char));
        for (int j = 0; j < size; j++) {
            board[i][j] = '_';
        }
    }
    return board;
}

void freeBoard(char **board, int size) {
    for (int i = 0; i < size; i++) {
        free(board[i]);
    }
    free(board);
}

bool notContiguous(int l, int c, char **map) {
    int directions[8][2] = { {-1, 0}, {1, 0}, {0, -1}, {0, 1},
                             {-1, -1}, {-1, 1}, {1, -1}, {1, 1} };
    
    for (int i = 0; i < 8; i++) {
        int dr = directions[i][0];
        int dc = directions[i][1];
        if (l + dr >= 0 && l + dr < mapSize && c + dc >= 0 && c + dc < mapSize) {
            if (map[l + dr][c + dc] == 'b') {
                return false;
            }
        }
    }
    return true;
}

bool isValid(int l, int c, char **map, char pistoleiro, char enemy) {
    int enemySeen = 0;
    int directions[8][2] = { {-1, 0}, {1, 0}, {0, -1}, {0, 1},
                             {-1, -1}, {-1, 1}, {1, -1}, {1, 1} };
    
    for (int i = 0; i < 8; i++) {
        int dr = directions[i][0];
        int dc = directions[i][1];
        for (int step = 1; step < mapSize; step++) {
            int r = l + dr * step;
            int d = c + dc * step;
            if (r < 0 || r >= mapSize || d < 0 || d >= mapSize) {
                break;
            }
            if (map[r][d] == pistoleiro) {
                return false;
            }
            if (map[r][d] == enemy) {
                enemySeen++;
                break;
            }
        }
    }
    return enemySeen >= 2;
}

bool checaTabuleiro(char **map) {
    for (int i = 0; i < mapSize; i++) {
        for (int j = 0; j < mapSize; j++) {
            if (map[i][j] == 'b') {
                if (!isValid(i, j, map, 'b', 'c')) {
                    return false;
                }
            }
        }
    }
    return true;
}

int salao(int b, int c, char **map, int startB, int startC) {
    int sum = 0;
    if (b == 0 && c == 0) {
        if (checaTabuleiro(map)) {
            return 1;
        }
        return 0;
    }

    if (b < 0 || c < 0) {
        return 0;
    }

    int size = mapSize;
    if (b > 0) {
        for (int z = startB; z < size * size; z++) {
            int i = z / size;
            int j = z % size;
            if (map[i][j] == '_' && notContiguous(i, j, map)) {
                map[i][j] = 'b';
                sum += salao(b - 1, c, map, z + 1, startC);
                map[i][j] = '_';
            }
        }
    } else if (c > 0) {
        for (int z = startC; z < size * size; z++) {
            int i = z / size;
            int j = z % size;
            if (map[i][j] == '_' && isValid(i, j, map, 'c', 'b')) {
                map[i][j] = 'c';
                sum += salao(b, c - 1, map, startB, z + 1);
                map[i][j] = '_';
            }
        }
    }

    return sum;
}

int main(int argc, char *argv[]) {
    if (argc < 4) {
        printf("Usage: %s mapSize b c\n", argv[0]);
        return 1;
    }

    mapSize = atoi(argv[1]);
    b = atoi(argv[2]);
    c = atoi(argv[3]);

    char **tabuleiro = createBoard(mapSize);

    clock_t inicio = clock();
    int solucoes = salao(b, c, tabuleiro, 0, 0);
    clock_t fim = clock();

    double tempo_execucao = (double)(fim - inicio) / CLOCKS_PER_SEC;
    printf("Soluções: %d\n", solucoes);
    printf("Tempo de execução: %.6f segundos\n", tempo_execucao);

    freeBoard(tabuleiro, mapSize);
    return 0;
}
